### PR TITLE
[Windows] Add a `flutter-dev` script to run the flutter CLI from source

### PR DESCRIPTION
**Pull request by loic-sharma**
Follow-up to https://github.com/flutter/flutter/pull/153599, adds support for the objectively better OS 👼

### CODE REVIEW

**Review by christopherfujino — APPROVED** [review]
i'm gonna take your word for it that this works :)

File: bin/flutter-dev [hunk 1]
```diff
@@ -50,6 +58,12 @@ PROG_NAME="$(follow_links "${BASH_SOURCE[0]}")"
 BIN_DIR="$(cd "${PROG_NAME%/*}" ; pwd -P)"
 SHARED_NAME="$BIN_DIR/internal/shared.sh"
 FLUTTER_ROOT="$(cd "$BIN_DIR/.." ; pwd -P)"
+OS="$(uname -s)"
+
+# If we're on Windows, invoke the batch script instead to get proper locking.
+if [[ $OS =~ MINGW.* || $OS =~ CYGWIN.* || $OS =~ MSYS.* ]]; then
+  exec "${BIN_DIR}/flutter-dev.bat" "$@"
+fi
```
Comment (loic-sharma) [inline bin/flutter-dev:66]: This makes `bin/flutter-dev` consistent with `bin/flutter`

File: bin/flutter-dev.bat [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+REM Chaining the call to 'dart' and 'exit' with an ampersand ensures that
+REM Windows reads both commands into memory once before executing them. This
+REM avoids nasty errors that may otherwise occur when the dart command (e.g. as
+REM part of 'flutter upgrade') modifies this batch script while it is executing.
+REM
+REM Do not use the CALL command in the next line to execute Dart. CALL causes
+REM Windows to re-read the line from disk after the CALL command has finished
+REM regardless of the ampersand chain.
+"%dart%" --packages="%flutter_tools_dir%\.dart_tool\package_config.json" %FLUTTER_TOOL_ARGS% "%script_path%" %* & "%exit_with_errorlevel%"
```
Comment (loic-sharma) [inline bin/flutter-dev.bat:60]: This line is different from `bin/flutter.bat`

File: bin/flutter-dev.bat [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  EXIT 1
+)
+
+REM Include shared scripts in shared.bat
+SET shared_bin=%FLUTTER_ROOT%\bin\internal\shared.bat
+CALL "%shared_bin%"
+
+SET flutter_tools_dir=%FLUTTER_ROOT%\packages\flutter_tools
+SET cache_dir=%FLUTTER_ROOT%\bin\cache
+SET script_path=%flutter_tools_dir%\bin\flutter_tools.dart
```
Comment (loic-sharma) [inline bin/flutter-dev.bat:46]: This line is different from `bin/flutter.bat`

**Comment by loic-sharma** [comment]
@christopherfujino Would you have any recommendations on tests I could add for `bin/flutter-dev.bat`? Or would you recommend I get a text exemption instead?

**Comment by matanlurey** [comment]
@loic-sharma: Thanks!

Let's also add the flag that Slava added here:
https://github.com/flutter/flutter/pull/153905

**Review by matanlurey — APPROVED** [review]
test-exempt: This is a dev-only tool, we don't mandate testing of these.

**Comment by loic-sharma** [comment]
> Let's also add the flag that Slava added here: #153905

Wow that's a massive improvement, that's awesome. Added!

**Pull request merged**